project_euler/problem_191/sol1.py

```diff
@@ -0,0 +1,105 @@
+"""
+Prize Strings
+Problem 191
+
+A particular school offers cash rewards to children with good attendance and
+punctuality. If they are absent for three consecutive days or late on more
+than one occasion then they forfeit their prize.
+
+During an n-day period a trinary string is formed for each child consisting
+of L's (late), O's (on time), and A's (absent).
+
+Although there are eighty-one trinary strings for a 4-day period that can be
+formed, exactly forty-three strings would lead to a prize:
+
+OOOO OOOA OOOL OOAO OOAA OOAL OOLO OOLA OAOO OAOA
+OAOL OAAO OAAL OALO OALA OLOO OLOA OLAO OLAA AOOO
+AOOA AOOL AOAO AOAA AOAL AOLO AOLA AAOO AAOA AAOL
+AALO AALA ALOO ALOA ALAO ALAA LOOO LOOA LOAO LOAA
+LAOO LAOA LAAO
+
+How many "prize" strings exist over a 30-day period?
+
+References:
+    - The original Project Euler project page:
+      https://projecteuler.net/problem=191
+"""
+
+
+cache = {}
+
+
+def _calculate(days: int, absent: int, late: int) -> int:
+    """
+    A small helper function for the recursion, mainly to have
+    a clean interface for the solution() function below.
+
+    It should get called with the number of days (corresponding
+    to the desired length of the 'prize strings'), and the
+    initial values for the number of consecutive absent days and
+    number of total late days.
+
+    >>> _calculate(days=4, absent=0, late=0)
+    43
+    >>> _calculate(days=30, absent=2, late=0)
+    0
+    >>> _calculate(days=30, absent=1, late=0)
+    98950096
+    """
+
+    # if we are absent twice, or late 3 consecutive days,
+    # no further prize strings are possible
+    if late == 3 or absent == 2:
+        return 0
```

Comment:
> I cannot hold your hand throughout your life :)

I will help you out here with this code as an example. The code here says if the `late` variable is 3 or the `absent` variable is 2 then we will `return` 0 and the comment says the same thing. What I want you to see is that comments are meant to explain the code **ONLY** if it's necessary but here the code is pretty self-explanatory. The comment here says the same thing in English.

What I basically mean here is that your code is perfect, you shouldn't need a ton of comments to explain what you have written. You don't want to keep referring to a dictionary when you're reading a book otherwise it breaks your flow.